core/src/main/scala/org/http4s/Headers.scala

```diff
@@ -1,158 +1,199 @@
-    * @return a new [[Headers]] containing the sum of the initial and input headers
-    */
-  def put(in: Header*): Headers =
-    if (in.isEmpty) this
-    else if (this.isEmpty) new Headers(in.toList)
-    else this ++ in
-
-  /** Concatenate the two collections
-    * If the resulting collection is of Headers type, duplicate Singleton headers will be removed from
-    * this Headers collection.
-    *
-    * @param that collection to append
-    * @tparam B type contained in collection `that`
-    * @tparam That resulting type of the new collection
-    */
-  override def ++[B >: Header, That](that: GenTraversableOnce[B])(
-      implicit bf: CanBuildFrom[Headers, B, That]): That =
-    if (bf eq Headers.canBuildFrom) { // Making a new Headers collection from a collection of Header's
-      if (that.isEmpty) this.asInstanceOf[That]
-      else if (this.isEmpty) that match {
-        case hs: Headers => hs.asInstanceOf[That]
-        case hs => new Headers(hs.toList.asInstanceOf[List[Header]]).asInstanceOf[That]
-      } else {
-        val hs = that.toList.asInstanceOf[List[Header]]
-        val acc = new ListBuffer[Header]
-        this.headers.foreach { orig =>
-          orig.parsed match {
-            case _: Header.Recurring => acc += orig
-            case _: `Set-Cookie` => acc += orig
-            case h if !hs.exists(_.name == h.name) => acc += orig
-            case _ => // NOOP, drop non recurring header that already exists
-          }
-        }
-
-        val h = new Headers(acc.prependToList(hs))
-        h.asInstanceOf[That]
-      }
-    } else super.++(that)
-
-  override def hashCode(): Int = this.headers.hashCode()
-
-  override def equals(that: Any): Boolean = that match {
-    case otherheaders: Headers =>
-      this.toList.equals(otherheaders.toList)
-    case _ => false
+// import scala.collection.{GenTraversableOnce, immutable, mutable}
+// import scala.collection.generic.CanBuildFrom
+// import scala.collection.mutable.ListBuffer
+
+
+final case class Headers(raw: List[Header.RawHeader]){
+  def get[A](implicit A: Header[A]): Option[A] = {
+    if (A.isSingleton) {
+      raw.find(_.name == A.name).flatMap(h => A.parse(h.value))
+    } else {
+      val values = raw.filter(_.name == A.name).map(_.value)
+      if (values.nonEmpty) A.parse(values.mkString(", "))
+      else None
+    }
   }
 
-  /** Removes the `Content-Length`, `Content-Range`, `Trailer`, and
-    * `Transfer-Encoding` headers.
-    *
-    *  https://tools.ietf.org/html/rfc7231#section-3.3
-    */
-  def removePayloadHeaders: Headers =
-    filterNot(h => Headers.PayloadHeaderKeys(h.name))
-
-  def redactSensitive(
-      redactWhen: CaseInsensitiveString => Boolean = Headers.SensitiveHeaders.contains): Headers =
-    headers.map {
-      case h if redactWhen(h.name) => Header.Raw(h.name, "<REDACTED>")
-      case h => h
+  def getRepeatable(key: CaseInsensitiveString): Option[Header.RawHeader] = {
```

Comment:
Why wouldn't this return a `List`?  Then it would be safe to call with `"Set-Cookie".ci `?

Reply:
Hmmm. Yeah.